TestingLowerBounds/KullbackLeibler.lean

```diff
@@ -354,46 +451,37 @@ lemma kl_compProd [CountablyGenerated β] [IsMarkovKernel κ] [IsMarkovKernel η
       filter_upwards [h] with x hx
       rw [hx, llr_def]
     apply integral_congr_ae
-    filter_upwards
-    intro a
+    filter_upwards with a
     congr
-  _ = ↑(∫ (a : α), log (μ.rnDeriv ν a).toReal ∂μ)
-      + ↑(∫ (a : α), ∫ (x : β), log (kernel.rnDeriv κ η a x).toReal ∂κ a ∂μ) := by
+  _ = ∫ a, log (μ.rnDeriv ν a).toReal ∂μ
+      + ∫ a, ∫ x, log ((κ a).rnDeriv (η a) x).toReal ∂κ a ∂μ := by
     simp only [integral_const, measure_univ, ENNReal.one_toReal, smul_eq_mul, one_mul]
-  _ = ↑(∫ (a : α), log (μ.rnDeriv ν a).toReal ∂μ)
-      + ↑(∫ (a : α), ∫ (x : β), log ((κ a).rnDeriv (η a) x).toReal ∂κ a ∂μ) := by
     congr 2
-    apply integral_congr_ae
+    apply integral_congr_ae₂
     filter_upwards [hκη] with a ha
     have h := ha.ae_le (kernel.rnDeriv_eq_rnDeriv_measure κ η a)
-    apply integral_congr_ae
     filter_upwards [h] with x hx
     congr
   _ = kl μ ν + condKL κ η μ := by
-    congr
-    · rw [← llr_def, ← kl_of_ac_of_integrable hμν]
-      exact intμν
-    · simp_rw [← llr_def]
-      rw [condKL_of_ae_ne_top_of_integrable _ _]
-      rotate_left
-      · filter_upwards [hκη, intκη2] with a ha hκηa
-        intro h
-        apply kl_eq_top_iff.mp at h
-        tauto
-      · apply Integrable.congr intκη
-        filter_upwards [hκη, intκη2] with a ha hκηa
-        rw [kl_of_ac_of_integrable ha hκηa, EReal.toReal_coe]
-        apply integral_congr_ae
-        filter_upwards [ha.ae_le (kernel.rnDeriv_eq_rnDeriv_measure κ η a)] with x hx
-        rw [hx, llr_def]
-      norm_cast
-      apply integral_congr_ae
+    congr <;> simp_rw [← llr_def]
+    · rw [← kl_of_ac_of_integrable hμν intμν]
+    · rw [condKL_of_ae_ac_of_ae_integrable_of_integrable' hκη intκη2 _]
+      apply (integrable_kl_iff hκη intκη2).mpr
+      simp_rw [llr_def]
+      apply Integrable.congr intκη
       filter_upwards [hκη] with a ha
-      by_cases h : Integrable (llr (κ a) (η a)) (κ a)
-      · suffices hh : kl (κ a) (η a) = ∫ x, llr (κ a) (η a) x ∂(κ a) from by simp [hh]
-        exact kl_of_ac_of_integrable (ha) h
-      · rw [kl_of_not_integrable h]
-        simp only [h, not_false_eq_true, integral_undef, EReal.toReal_top]
+      have h := ha.ae_le (kernel.rnDeriv_eq_rnDeriv_measure κ η a)
+      apply integral_congr_ae
+      filter_upwards [h] with x hx
+      rw [hx]
+
+--TODO: decide the name for this lemma, in the blueprint it is called kl_chain_rule_prod, but if we call it like that maybe we have to change also the name of the previous one. A possible name could be kl_joint, but I'm not sure about it
+--Why do we need to assume that β is not empty?
```

Comment:
If I remember correctly, beta needs to be nonempty for condKernel to exist, otherwise we can't say that there exists a suitable function.